pkg/operator/controller/dns_status.go

```diff
@@ -92,34 +173,137 @@ func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clu
+			})
 		case have == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsAvailable")
-			messages = append(messages, "No DNS pods are available.")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{
+				Type:    DNSNoDNSPodsAvailable,
+				Reason:  DNSNoDNSPodsAvailable,
+				Status:  operatorv1.ConditionTrue,
+				Message: "No DNS pods are available.",
+			})
 		case intstrErr != nil:
-			degradedReasons = append(degradedReasons, "InvalidDNSMaxUnavailable")
-			messages = append(messages, fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr))
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{
+				Type:    DNSInvalidDNSMaxUnavailable,
+				Reason:  DNSInvalidDNSMaxUnavailable,
+				Status:  operatorv1.ConditionTrue,
+				Message: fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr),
+			})
 		case int(numberUnavailable) > maxUnavailable:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "MaxUnavailableDNSPodsExceeded")
-			messages = append(messages, fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable))
+			graceCondition = &operatorv1.OperatorCondition{
+				Type:    DNSMaxUnavailableDNSPodsExceeded,
+				Reason:  DNSMaxUnavailableDNSPodsExceeded,
+				Status:  operatorv1.ConditionTrue,
+				Message: fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable),
+			}
 		}
 	}
 
-	if len(degradedReasons) != 0 {
-		degradedCondition.Status = status
-		degradedCondition.Reason = strings.Join(degradedReasons, "")
-		degradedCondition.Message = strings.Join(messages, "\n")
+	// If there are degraded conditions, there is no grace period, and this should not be requeued.
+	// Set the final degraded condition as true, and concatenate the reasons and messages
+	if len(degradedConditions) != 0 {
+		reasons, messages := formatConditions(degradedConditions)
+		degradedCondition.Status = operatorv1.ConditionTrue
+		degradedCondition.Reason = reasons
+		degradedCondition.Message = messages
+		degradedConditions = append(degradedConditions, setDNSLastTransitionTime(degradedCondition, oldDegradedCondition))
+
+		return degradedConditions, nil
+	}
+
+	// If the graceCondition occurs, this should be requeued.
+	// Set the final degraded condition as false, add the grace condition, and set it up to requeue the reconcile.
+	// We check this last, because the other conditions are non-recoverable as degraded conditions.
+	if graceCondition.Status == operatorv1.ConditionTrue {
+		requeueTime := requeueAfter(expectedConditions, graceCondition, oldGraceCondition)
+		if requeueTime == 0 {
+			graceCondition.Status = operatorv1.ConditionUnknown
+		}
+		degradedConditions = append(degradedConditions, setDNSLastTransitionTime(graceCondition, oldGraceCondition))
```

Comment:
Similar question to https://github.com/openshift/cluster-dns-operator/pull/290#discussion_r742436545:  Do we only persist the `MaxUnavailableDNSPodsExceeded` status condition if its status is `True`?  Once it's been set to `True`, does anything ever delete it or update it to `False`?